check_test.go

```diff
@@ -167,6 +167,50 @@ func main() {
 	assert.Len(t, errs, 0)
 }
 
+// TestNoMissingInterface tests that we correctly detect exhaustive case when
+// there is an interface which implements the interface we are going to check.
+func TestNoMissingInterface(t *testing.T) {
+	code := `
+package main
+
+//go-sumtype:decl T
+
+type T interface {
+	sealedT()
+}
+
+type A struct {}
+func (a *A) sealedT() {}
+
+type U interface {
+	T
+	sealedU()
+}
+
+type B struct {}
+func (b *B) sealedT() {}
+func (b *B) sealedU() {}
+
+type C struct {}
+func (c *C) sealedT() {}
+func (c *C) sealedU() {}
+
+func main() {
+	switch T(nil).(type) {
+	case *A, *B, *C:
+	}
+	switch T(nil).(type) {
+	case *A, U:
+	}
```

Comment:
So this test case is somewhat interesting. Your argument is compelling, but I don't think I'm quite 100% convinced yet. The problem, as I see it, is that _both_ of these type switch statements, when used in real code, could execute without a single case firing. For example, in the former case, if `x.(type)` is a `U`, then none of the `*A`, `*B` or `*C` cases will catch it. Similarly, in the latter case, if `x.(type)` is a `*B` or a `*C`, then neither the `*A` or `U` cases will catch it. So basically, what you wind up with is a type switch that is not actually exhaustive. To me, this seems like a fundamental violation of what this tool is trying to do. That is, if this tool passes, then your switch statements should _either_ be exhaustive for all sum types, or explicitly _non-exhaustive_ with a `default` case.

On the other hand, I do agree that writing code that handles all of `*A`, `*B`, `*C` and `U` is likely not common and probably not what you want. But this change overall weakens what "exhaustiveness" actually means. So I'm not sure what to do. :-/

Reply:
> So this test case is somewhat interesting. Your argument is compelling, but I don't think I'm quite 100% convinced yet. The problem, as I see it, is that _both_ of these type switch statements, when used in real code, could execute without a single case firing.

Since `T` is sealed, `U` is also sealed if it implements `T`. All concrete types implementing `U` must also be defined in the same package. Therefore, `U` is a sum type, and `U` is a subset of `T`.

> For example, in the former case, if `x.(type)` is a `U`, then none of the `*A`, `*B` or `*C` cases will catch it.

If `U` implements `T`, a `U` is either `*B`, `*C`, or `nil`. The only missing case here is `nil` and listing `U` again does not help.

If `U` does not implement `T`, the type switch is still exhaustive because types implementing `U` but do not implementing `T` cannot be stored in a `T`. Therefore, listing `*B` and `*C` is sufficient.

> Similarly, in the latter case, if `x.(type)` is a `*B` or a `*C`, then neither the `*A` or `U` cases will catch it.

Since `*B` and `*C` implement `U`, `*B` or `*C` must match `U` regardless of whether `U` implements `T`.

> So basically, what you wind up with is a type switch that is not actually exhaustive. To me, this seems like a fundamental violation of what this tool is trying to do. That is, if this tool passes, then your switch statements should _either_ be exhaustive for all sum types, or explicitly _non-exhaustive_ with a `default` case.

I think a type switch is still exhaustive in both cases, but I agree it may look confusing if `U` does not implement `T`. Do you think it will look better if we require `U` to implement `T`?

Reply:
Ping ...